tools/contractdeployer/obscuro_deployer.go

```diff
@@ -0,0 +1,127 @@
+	err = fundDeployerWithFaucet(cfg, client)
+	if err != nil {
+		return nil, fmt.Errorf("failed to fund deployer acc from faucet - %w", err)
+	}
+
+	return &obscuroDeployer{client: client}, nil
+}
+
+func fundDeployerWithFaucet(cfg *Config, client *obsclient.AuthObsClient) error {
+	// Create the L2 faucet wallet and client.
+	faucetPrivKey, err := crypto.HexToECDSA(rollupchain.FaucetPrivateKeyHex)
+	if err != nil {
+		panic("could not initialise L2 faucet private key")
+	}
+	faucetWallet := wallet.NewInMemoryWalletFromPK(cfg.ChainID, faucetPrivKey)
+
+	faucetClient, err := connectClient(getURL(cfg), faucetWallet)
+	if err != nil {
+		return err
+	}
+
+	balance, err := client.BalanceAt(context.TODO(), nil)
+	if err != nil {
+		return fmt.Errorf("failed to fetch contract deployer balance. Cause: %w", err)
+	}
+	// We do not send more funds if the contract deployer already has enough.
+	if balance.Cmp(big.NewInt(Prealloc)) == 1 {
+		return nil
+	}
+
+	nonce, err := faucetClient.NonceAt(context.TODO(), nil)
+	if err != nil {
+		return fmt.Errorf("failed to fetch faucet nonce. Cause: %w", err)
+	}
+
+	destAddr := client.Address()
+	tx := &types.LegacyTx{
+		Nonce:    nonce,
+		Value:    big.NewInt(Prealloc),
+		Gas:      uint64(1_000_000),
+		GasPrice: gethcommon.Big1,
+		To:       &destAddr,
+	}
+	signedTx, err := faucetWallet.SignTransaction(tx)
+	if err != nil {
+		return fmt.Errorf("failed to sign faucet transaction: %w", err)
+	}
+
+	err = faucetClient.SendTransaction(context.TODO(), signedTx)
+	if err != nil {
+		return fmt.Errorf("failed to send contract deploy transaction: %w", err)
+	}
+
+	rec, err := clientutil.AwaitTransactionReceipt(context.TODO(), faucetClient, signedTx.Hash(), timeoutWait)
+	if err != nil {
+		return fmt.Errorf("failed to complete faucet transaction. Cause: %w", err)
+	}
+	if rec.Status != types.ReceiptStatusSuccessful {
+		return fmt.Errorf("faucet transaction status unsuccessful: %v", rec)
+	}
+
+	return nil
+}
+
+func connectClient(url string, wal wallet.Wallet) (*obsclient.AuthObsClient, error) {
+	var client *obsclient.AuthObsClient
+	var err error
+
+	startConnectingTime := time.Now()
+	// since the nodes we are connecting to may have only just started, we retry connection until it is successful
+	for client == nil && time.Since(startConnectingTime) < timeoutWait {
+		client, err = obsclient.DialWithAuth(url, wal)
+		if err == nil {
+			break // success
+		}
+		// if there was an error we'll retry, if we timeout the last seen error will display
+		time.Sleep(retryInterval)
+	}
+	if client == nil {
+		return nil, fmt.Errorf("failed to initialise client connection after retrying for %s, %w", timeoutWait, err)
+	}
+	return client, nil
+}
+
+type obscuroDeployer struct {
+	client *obsclient.AuthObsClient
+}
+
+func (o *obscuroDeployer) Nonce(_ gethcommon.Address) (uint64, error) {
+	return o.client.NonceAt(context.TODO(), nil)
```

Comment:
let's add a check here to make sure the address passed in is the same as the one in the client.
to prevent weird errors